### PR TITLE
JieJhih proposal draft1

### PR DESCRIPTION
**Pull request by LexCC**
---
A good PR should include the following parts:

- [x] A clear title
- [x] A meaningful message for PR, as well as its commits
- [x] from your specific branch (***not master***) merging to teacher's `HW` branch (***not master***, again)
- [x] excluding any irrelevant files, such as binaries, text files, or dot files
- [ ] passing tests/CI

### CODE REVIEW

**Comment by yungyuc** [comment]
@JieJhih , I am not finding this PR in the e3 entry and not aware of your GitHub account.  As a result I do not know who you are.  Before you provide me the requested information, I cannot grade your proposal.

**Comment by LexCC** [comment]
@yungyuc Sorry I forgot to paste the link of PR in the e3. The "nsd term project proposal" is closed in the e3, could you reopen it or should I just provide the requested information here? I apologize for any inconvenience.

**Comment by yungyuc** [comment]
@JieJhih It's fine to have the e3 submission missing, but you may have some points deducted.  But please follow the directive in the e3 course announcement and send me the email for the information I requested.  Without association to your e3 account, you may not receive any points.

File: proposal/JieJhih/README.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+==========================
+Packet Analyzer is a tool to capture the packets and make it visualize.
+To capture the packets I use C++ to write a eBPF program, which makes the speed of capture packets is really fast. (Please follow [Packet flow in Netfilter and General Networking](https://upload.wikimedia.org/wikipedia/commons/3/37/Netfilter-packet-flow.svg)) 
+After get a bunch of packets, we use python to visualize those packets.
+
+Basic Information
+=================
+
+GitHub repository:
+https://github.com/JieJhih/Packet-Analyzer.git
```
Comment (yungyuc) [inline proposal/JieJhih/README.md:11]: I hope you make some progress soon.  It's the eighth week and you have less than 2 months to go.

File: proposal/JieJhih/README.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+Problem To Solve
+================
+
+A server will receive a bunch of packets every day, sometimes system administrator need to analyze each kind of packet (TCP? UDP? Specific IP?) occupy how many percent in daily. Using eBPF program is faster than other userspace tools, cooperate with Python could make data visible.
```
Comment (yungyuc) [inline proposal/JieJhih/README.md:16]: How many data do you expect the system to process or analyze?

File: proposal/JieJhih/README.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+Perspective users
+=================
+1. System administrator who want to maintain server's health.
+2. Data scientists who want to analyze the client agent, and packets peak or off-peak hours.
+
+System architecture
+=================
+
+1. Capture the packets from eBPF program.
+2. Send packets information to Python.
```
Comment (yungyuc) [inline proposal/JieJhih/README.md:27]: If you send each of the packets to Python, the system will run quite slowly.  A more performant design is to have a container of the packet information in C++ and Python only access it through handles.

File: proposal/JieJhih/README.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+3. Organize a bunch of information of packets.
+4. Show the statistic using Python GUI tool.
```
Comment (yungyuc) [inline proposal/JieJhih/README.md:29]: The GUI part will not be considered in grading.

File: proposal/JieJhih/README.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+API decription
+=================
+1. Packet pool buffers.
```
Comment (yungyuc) [inline proposal/JieJhih/README.md:33]: I think it makes sense to pool the (packet) information.  But this seems to contradict to your second bullet in System architecture.

File: proposal/JieJhih/README.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+Engineering infrastructure
+=================
+1. Basic packet syntax parser
+2. Basic packet data structure
+3. Packets record usage
+
+Schedule
+=================
+
+- week 7 written eBPF program in C++
```
Comment (yungyuc) [inline proposal/JieJhih/README.md:47]: I think you need to add testing and CI to the plan.

**Review by yungyuc — APPROVED** [review]

**Pull request merged**